nengo_gui/components/value.py

```diff
@@ -59,6 +59,9 @@ def update_client(self, client):
         # while there is data that should be sent to the client
         while len(self.data) > 0:
             item = self.data.popleft()
+            while len(self.data) > 0:
```

Comment:
> In that case it might be possible, that data is added after the inner while loop finished, but before the condition of the outer while loop is checked. In that case the code would enter the outer while loop again, overwriting item and thus it loses some data.

That's the whole point of the nested while loops.  It's supposed to go enter the outer while loop again if new data is added concurrently after the inner while loop finishes.  And if that does happen, I want ```item``` to be overwritten because it has already been sent to the client.

I think part of the problem here is that a) it looks weird to have two nested while loops like that, and b) the blank line after the inner while loop is leading people to think that's also the end of the outer while loop (i.e. it's easy to not notice that the ```client.write``` line is inside the outer while loop).  

So I'm still pretty sure that the code is algorithmically correct (and doesn't have any race conditions), but the fact that two different people found it confusing and weird definitely means I should rewrite it.  And I think something like replacing the inner loop with ```''.join(self.data[:length])``` should do it and be more efficient anyway (and be thread-safe).